mautic/mautic
# Replacing HistoryJS with native history browser api

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [y]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

~~⚠️ this PR is built on top of https://github.com/mautic/mautic/pull/12374 so that should be merged first.~~

The [HistoryJS](https://www.npmjs.com/package/historyjs?activeTab=readme) library was updated 10 years ago. It has a dependency on a `Buildr` JS library that should have been a dev-dependency IMHO but it's not. So when we install NPM dependencies we get this warning:
```
10 vulnerabilities (1 low, 4 high, 5 critical)
```
It turned out all of these were reported in the Buildr dependencies. So when we remove the HistoryJS library from Mautic then we get clean sheet:
```
found 0 vulnerabilities
```
I suppose the reason that the HistoryJS library was not updated in 10 years and there are no replacement libraries is that the native [History browser API is stable enough](https://caniuse.com/?search=history) then there is no need for a specific library.

I tried that by using these 2 native browser APIs instead of the HistoryJS polyfil:
https://developer.mozilla.org/en-US/docs/Web/API/History_API
https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event

And it seem to be that it's working just fine.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Click through the UI. The content is being loaded via JS. There are no refreshes of the whole page.
3. Click the browser's back button. You should be redirected to the previous page. Not the first page you hard-reloaded.
4. Click the browser's forward button. You should be redirected back to the previous page before you clicked the back button.
